src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Razor/ExtractToNewComponentCodeActionProvider.cs

```diff
@@ -100,4 +160,64 @@ private static bool TryGetNamespace(RazorCodeDocument codeDocument, [NotNullWhen
         // and causing compiler errors. Avoid offering this refactoring if we can't accurately get a
         // good namespace to extract to
         => codeDocument.TryComputeNamespace(fallbackToRootNamespace: true, out @namespace);
+
+    public (SyntaxNode? Start, SyntaxNode? End) FindContainingSiblingPair(SyntaxNode startNode, SyntaxNode endNode)
+    {
+        // Find the lowest common ancestor of both nodes
+        var lowestCommonAncestor = FindLowestCommonAncestor(startNode, endNode);
+        if (lowestCommonAncestor == null)
+        {
+            return (null, null);
+        }
+
+        SyntaxNode? startContainingNode = null;
+        SyntaxNode? endContainingNode = null;
+
+        // Pre-calculate the spans for comparison
+        var startSpan = startNode.Span;
+        var endSpan = endNode.Span;
+
+        foreach (var child in lowestCommonAncestor.ChildNodes().Where(node => node.Kind == SyntaxKind.MarkupElement))
```

Comment:
Given this starts from a common ancestor, doesn't that mean by definition it will contain both the start and end spans, and therefore this loop will only ever run one iteration, and always return both `endContainingNode` and `startContaininedNode` as being the same node as `lowestCommonAncestor`?

I might be missing something though. Again, so illustrative comments would be helpful.

Reply:
The childNodes method won't include the ancestor itself AFAIK, but on inspecting the code there's no need to iterate through all of the childNodes; the loop only ever iterates exactly 3 times.

Reply:
> The ChildNodes() method won't include the ancestor itself AFAIK.

Derp, of course 🤦‍♂️

> The loop runs exactly 3 iterations

Sounds like more test cases might help there too. If you can confidently say that the first iteration finds the start node, and the second finds the end node, it sounds like a very specific case where both nodes are directly next to each other. What would happen in this situation?

```
<common>
	[|<start>
	   </start>
	 <middle>
	</middle>
	<end>
	</end>|]
</common>
```

At least by me typing that out, I now understand more what the logic is trying to do :)